aten/src/ATen/native/sparse/mps/SparseMPSTensorMath.mm

```diff
@@ -1109,6 +1113,129 @@ Tensor sparse_sparse_matmul_mps(const Tensor& mat1_, const Tensor& mat2_) {
+
+    if (sum_dense_dim) {
+      auto dense_expand_size = std::vector<int64_t>(expand_size.begin() + 1, expand_size.end());
+      for (auto d : dense_dims_to_sum_v) {
+        grad_input_values = grad_input_values.unsqueeze(d - 1);
+      }
+      grad_input_values = grad_input_values.expand(dense_expand_size);
+    }
+    grad_input_values = grad_input_values.expand(expand_size).clone(at::MemoryFormat::Contiguous);
+
+    return at::_sparse_coo_tensor_with_dims_and_tensors(
+        input_sparse_dim, input_dense_dim, input_sizes,
+        input_indices.clone(at::MemoryFormat::Contiguous),
+        grad_input_values, input.options().dtype(grad_.dtype()));
+  }
+
+  TORCH_CHECK(grad_.is_sparse(), "_sparse_sum_backward_mps: expected 'grad_' Tensor to be sparse when not all sparse dims are summed");
+  auto grad = grad_.coalesce();
+  auto grad_indices = grad._indices();
+  auto grad_values = grad._values();
+  const auto grad_sparse_dim = grad.sparse_dim();
+  const auto grad_nnz = grad._nnz();
+
+  auto grad_values_expand = grad_values;
+  if (sum_dense_dim) {
+    auto expand_size = input_values.sizes().vec();
+    expand_size[0] = grad_values.size(0);
+    for (auto d : dense_dims_to_sum_v) {
+      grad_values_expand = grad_values_expand.unsqueeze(d);
+    }
+    grad_values_expand = grad_values_expand.expand(expand_size).clone(at::MemoryFormat::Contiguous);
+  }
+
+  Tensor grad_input_values;
+  if (!sum_sparse_dim) {
+    grad_input_values = grad_values_expand;
+  } else {
+    auto grad_keys = at::sparse::flatten_indices(
+        grad_indices.contiguous(),
+        grad.sizes().slice(0, grad_sparse_dim)).to(at::kInt).contiguous();
+
+    std::vector<int64_t> sizes_keep;
+    sizes_keep.reserve(sparse_dims_to_keep_v.size());
+    for (auto d : sparse_dims_to_keep_v) {
+      sizes_keep.push_back(input_sizes[d]);
+    }
+
+    auto keep_idx = at::tensor(sparse_dims_to_keep_v, at::device(input_indices.device()).dtype(at::kLong));
+    auto input_indices_keep = input_indices.index_select(0, keep_idx).contiguous();
+    auto input_keys = at::sparse::flatten_indices(
+        input_indices_keep, IntArrayRef(sizes_keep)).to(at::kInt).contiguous();
+
+    auto idx_in_grad = searchsorted_mps(grad_keys, input_keys);
+    auto idx_in_grad_clamped = idx_in_grad.clamp_max(grad_nnz - 1);
+    auto matched_mask = grad_keys.index_select(0, idx_in_grad_clamped).eq(input_keys);
+
+    grad_input_values = grad_values_expand.index_select(0, idx_in_grad_clamped);
+
+    if (grad_input_values.dim() > 1) {
+      auto reshape_mask_size = std::vector<int64_t>(grad_input_values.dim(), 1);
+      reshape_mask_size[0] = input_nnz;
+      matched_mask = matched_mask.view(reshape_mask_size);
+    }
+    grad_input_values.masked_fill_(matched_mask.logical_not(), 0);
+  }
+
+  return at::_sparse_coo_tensor_with_dims_and_tensors(
+      input_sparse_dim, input_dense_dim, input_sizes,
+      input_indices.clone(at::MemoryFormat::Contiguous),
```

Comment:
Why does it need an explicit clone vs a to or contiguous call? Are the input args mutated here?